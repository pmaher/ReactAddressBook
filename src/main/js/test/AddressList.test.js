require('./helpers/enzyme-setup');
import React from 'react';
import {configure, mount} from 'enzyme';
import { AddressList, Address, EditAddressButton } from '../AddressList';
import axios from 'axios';
import renderer from 'react-test-renderer';
//the MemoryRouter is used to get around issues around React context
import { MemoryRouter } from 'react-router-dom';
jest.mock('axios');
    
describe('AddressList', () => {
    
    it('should make request to backend to retrieve all addresses', () => {
    	const response = { data: [
    		{
    			id: 1,
    			firstName: "James",
    			lastName: "Carter",
    			email: "James.Carter@hotmail.com",
    			phone: "(805) 111-9999",
    			address: "110 W. Liberty St.",
    			address2: "",
    			city: "Madison",
    			state: "WI",
    			zipcode: "53001"
    		},
    		{
    			id: 2,
    			firstName: "Helen",
    			lastName: "Leary",
    			email: "Helen.Leary@yahoo.com",
    			phone: "(805) 222-8888",
    			address: "638 Cardinal Ave.",
    			address2: "Apt. 2",
    			city: "Sun Prairie",
    			state: "WI",
    			zipcode: "53002"
    		}
    	]};
    	axios.get.mockResolvedValue(response);
    	
    	const wrapper = mount(<MemoryRouter><AddressList /></MemoryRouter>);

    	//queue this assertion callback behind the axios mock promise being resolved
    	setImmediate(() => {
        	//this update is needed since the enzyme wrapper is immutable as of v.3
        	wrapper.update();
        	expect(wrapper).toMatchSnapshot();
    	});
    });
    
    
    it('should display "no results" message when no addresses are found', () => {
    	const response = { data: [] };
    	axios.get.mockResolvedValue(response);
    	const wrapper = mount(<MemoryRouter><AddressList /></MemoryRouter>);

    	//queue this assertion callback behind the axios mock promise being resolved
    	setImmediate(() => {
        	//this update is needed since the enzyme wrapper is immutable as of v.3
        	wrapper.update();
        	expect(wrapper).toMatchSnapshot();
    	});
    });

});

describe('EditAddressButton', () => {
	it('should navigate to the url passed in via props when clicked', () => {
		const pushMock = jest.fn(),
			historyMock = { "push": pushMock },
			//Then we tell enzyme to explicitly only render the WrappedComponent, not the router, to access the mock history props
			wrapper = mount(<MemoryRouter><EditAddressButton.WrappedComponent to={"/edit/3"} history={historyMock}>Edit</EditAddressButton.WrappedComponent></MemoryRouter>);

		const button = wrapper.find('button').at(0);
		button.prop('onClick')();
		expect(pushMock.mock.calls.length).toBe(1);
		// The first argument of the first call to the function was 0
		expect(pushMock.mock.calls[0][0]).toBe("/edit/3");
	});
});

