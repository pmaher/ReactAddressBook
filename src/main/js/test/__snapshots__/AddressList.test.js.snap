// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressList should display "no results" message when no addresses are found 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <AddressList>
      <table className=\\"table table-striped table-bordered table-hover\\">
        <tbody>
          <tr>
            <th>
              First Name
            </th>
            <th>
              Last Name
            </th>
            <th>
              Email
            </th>
            <th>
              Phone
            </th>
            <th />
          </tr>
          <tr>
            <td colSpan=\\"5\\">
              No Results Found...
            </td>
          </tr>
        </tbody>
      </table>
    </AddressList>
  </Router>
</MemoryRouter>"
`;

exports[`AddressList should make request to backend to retrieve all addresses 1`] = `
"<MemoryRouter>
  <Router history={{...}}>
    <AddressList>
      <table className=\\"table table-striped table-bordered table-hover\\">
        <tbody>
          <tr>
            <th>
              First Name
            </th>
            <th>
              Last Name
            </th>
            <th>
              Email
            </th>
            <th>
              Phone
            </th>
            <th />
          </tr>
          <Address address={{...}}>
            <tr>
              <td>
                James
              </td>
              <td>
                Carter
              </td>
              <td>
                James.Carter@hotmail.com
              </td>
              <td>
                (805) 111-9999
              </td>
              <td>
                <withRouter() to=\\"/edit/1\\">
                  <Route>
                    <Component to=\\"/edit/1\\" match={{...}} location={{...}} history={{...}} staticContext={[undefined]}>
                      <button className=\\"btn btn-info\\" onClick={[Function: onClick]}>
                        Edit
                      </button>
                    </Component>
                  </Route>
                </withRouter()>
              </td>
            </tr>
          </Address>
          <Address address={{...}}>
            <tr>
              <td>
                Helen
              </td>
              <td>
                Leary
              </td>
              <td>
                Helen.Leary@yahoo.com
              </td>
              <td>
                (805) 222-8888
              </td>
              <td>
                <withRouter() to=\\"/edit/2\\">
                  <Route>
                    <Component to=\\"/edit/2\\" match={{...}} location={{...}} history={{...}} staticContext={[undefined]}>
                      <button className=\\"btn btn-info\\" onClick={[Function: onClick]}>
                        Edit
                      </button>
                    </Component>
                  </Route>
                </withRouter()>
              </td>
            </tr>
          </Address>
        </tbody>
      </table>
    </AddressList>
  </Router>
</MemoryRouter>"
`;
